tdx/measurement.go

```diff
@@ -0,0 +1,127 @@
+// • TDH.MNG.INIT begins the process by initializing the digest.
+// • TDH.MEM.PAGE.ADD adds a TD private page to the TD and inserts its properties (GPA) into the
+//   MRTD digest calculation.
+// • Control structure pages (TDR, TDCX, TDVPR and TDVPX) and Secure EPT pages are not measured.
+// • For pages whose data contribute to the TD, that data should be included in the TD measurement
+//   via TDH.MR.EXTEND. TDH.MR.EXTEND inserts the data contained in those pages and its GPA, in
+//   256-byte chunks, into the digest calculation. If a page will be wiped and initialized by TD
+//   code, the loader may opt not to measure the initial contents of the page with TDH.MR.EXTEND.
+// • The measurement is then completed by TDH.MR.FINALIZE. Once completed, further TDH.MEM.PAGE.ADDs
+//   or TDEXTENDs will fail.
+// MRTD extension by GPA uses a 128B buffer which includes the GPA and the leaf function name for
+// uniqueness.
+
+// Measurement represents the expected MRTD field of a TDX Quote.
+type Measurement struct {
+	digest hash.Hash
+}
+
+// NewMeasurement returns a new Measurement construct for calculating the TDX MRTD.
+func NewMeasurement() *Measurement {
+	return &Measurement{digest: sha512.New384()}
+}
+
+func (m *Measurement) extend(data []byte) {
+	m.digest.Write(data)
+}
+
+// TDH.MEM.PAGE.ADD [...]
+//  10. Extend TDCS.MRTD with the target page GPA. Extension is done using SHA384 with a 128B
+//     extension buffer composed as follows:
+//     o Bytes 0 through 11 contain the ASCII string “MEM.PAGE.ADD”.
+//     o Bytes 16 through 23 contain the GPA (in little-endian format).
+//     o All the other bytes contain 0.
+func (m *Measurement) pageAdd(gpa uint64) {
+	var buf [extensionBufferSize]byte
+	copy(buf[0:12], []byte("MEM.PAGE.ADD"))
+	binary.LittleEndian.PutUint64(buf[16:24], gpa)
+	m.extend(buf[:])
+}
+
+// TDH.MR.EXTEND
+// Extend TDCS.MRTD with the chunk’s GPA and contents. Extension is done using SHA384, with three
+// 128B extension buffers. The first extension buffer is composed as follows:
+// o Bytes 0 through 8 contain the ASCII string “MR.EXTEND”.
+// o Bytes 16 through 23 contain the GPA (in little-endian format).
+// o All the other bytes contain 0.
+// The other two extension buffers contain the chunk’s contents.
+func (m *Measurement) mrExtend(gpa uint64, data []byte) {
+	var buf [extensionBufferSize]byte
+	copy(buf[0:9], []byte("MR.EXTEND"))
+	binary.LittleEndian.PutUint64(buf[16:24], gpa)
+	m.extend(buf[:])
+	m.extend(data[0:extensionBufferSize])
+	m.extend(data[extensionBufferSize:])
+}
+
+// InitMemoryRegion extends a Measurement with the initial contents of a page.
+func (m *Measurement) InitMemoryRegion(region *ovmf.MaterialGuestPhysicalRegion) error {
+	gpr := region.GPR
+	data := region.HostBuffer
+	if gpr.Length != uint64(len(data)) {
+		return fmt.Errorf("gpr.Length %d does not match source data size %d", gpr.Length, len(data))
+	}
+	if gpr.Start%abi.PageSize != 0 {
```

Comment:
Can you add spaces around the % for readability?

Reply:
Nope. That's gofmt-enforced.